app/scripts/lib/segment/analytics.js

```diff
@@ -0,0 +1,264 @@
+    // Currently segment seems to support these source libraries only.
+    message.context = Object.assign({
+      library: {
+        name: 'analytics-node',
+      },
+    });
+
+    if (!message.timestamp) {
+      message.timestamp = new Date();
+    }
+
+    if (!message.messageId) {
+      message.messageId = generateRandomId();
+    }
+
+    if (message.anonymousId && !isString(message.anonymousId)) {
+      message.anonymousId = JSON.stringify(message.anonymousId);
+    }
+    if (message.userId && !isString(message.userId)) {
+      message.userId = JSON.stringify(message.userId);
+    }
+    this.queue.push({ message, callback });
+
+    if (!this.flushed) {
+      this.flushed = true;
+      this.flush();
+      return;
+    }
+
+    const hasReachedFlushAt = this.queue.length >= this.flushAt;
+    const hasReachedQueueSize =
+      this.queue.reduce((acc, item) => acc + JSON.stringify(item).length, 0) >=
+      this.maxQueueSize;
+    if (hasReachedFlushAt || hasReachedQueueSize) {
+      this.flush();
+    }
+
+    if (this.flushInterval && !this.timer) {
+      this.timer = setTimeout(this.flush.bind(this), this.flushInterval);
+    }
+  }
+
+  /**
+   * Flush the current queue
+   *
+   * @param {Function} [callback] - (optional)
+   * @returns {Analytics}
+   */
+  flush(callback = noop) {
+    if (!this.enable) {
+      setImmediate(callback);
+      return Promise.resolve();
+    }
+
+    if (this.timer) {
+      clearTimeout(this.timer);
+      this.timer = null;
+    }
+
+    if (!this.queue.length) {
+      setImmediate(callback);
+      return Promise.resolve();
+    }
+
+    const items = this.queue.splice(0, this.flushAt);
+    const callbacks = items.map((item) => item.callback);
+    const messages = items.map((item) => item.message);
+
+    const data = {
+      batch: messages,
+      timestamp: new Date(),
+      sentAt: new Date(),
+    };
+
+    const done = (err) => {
+      setImmediate(() => {
+        callbacks.forEach((fn) => fn(err, data));
+        callback(err, data);
+      });
+    };
+
+    const headers = {
+      Authorization: `Basic ${Buffer.from(this.writeKey, 'utf8').toString(
+        'base64',
+      )}`,
+    };
+
+    this._sendRequest(
+      `${this.host}${this.path}`,
+      {
+        method: 'POST',
+        body: JSON.stringify(data),
+        headers,
+      },
+      done,
+      0,
+    );
+  }
+
+  _sendRequest(url, body, callback, retryNo) {
+    fetch(url, body)
+      .then(async (res) => {
+        const response = await res.json();
+        if (res.ok) {
+          callback();
+        } else {
+          const error = new Error(res.statusText);
+          callback(error);
+        }
+        return Promise.resolve(response);
+      })
+      .catch((error) => {
+        if (this._isErrorRetryable(error) && retryNo <= this.retryCount) {
+          this._sendRequest(url, body, callback, retryNo + 1);
+        }
+        callback(error);
+      });
+  }
```

Comment:
I am finding it hard to test retry, plz share suggesting for QA.

Reply:
I'm using the AutoResponder feature of [Fiddler EveryWhere](https://www.telerik.com/download/fiddler-everywhere). Let me know if you would like a walkthrough